src/Headless/Avalonia.Headless.NUnit/AvaloniaTestMethodCommand.cs

```diff
@@ -45,22 +50,75 @@ private AvaloniaTestMethodCommand(
 
     public static TestCommand ProcessCommand(HeadlessUnitTestSession session, TestCommand command)
     {
-        return ProcessCommand(session, command, new List<Action>(), new List<Action>());
+        return ProcessCommand(session, command, [], []);
     }
 
-    private static TestCommand ProcessCommand(HeadlessUnitTestSession session, TestCommand command, List<Action> before, List<Action> after)
+    private static TestCommand ProcessCommand(HeadlessUnitTestSession session, TestCommand command, List<Action> before, List<Action<TestExecutionContext>> after)
     {
         if (command is BeforeAndAfterTestCommand beforeAndAfterTestCommand)
         {
             if (s_beforeTest.GetValue(beforeAndAfterTestCommand) is Action<TestExecutionContext> beforeTest)
             {
-                Action<TestExecutionContext> beforeAction = c => before.Add(() => beforeTest(c));
-                s_beforeTest.SetValue(beforeAndAfterTestCommand, beforeAction);
+                var setUpTearDownInfo = beforeTest.Target?.GetType()
+                    .GetField("setUpTearDown", BindingFlags.Instance | BindingFlags.Public);
```

Comment:
I don't know how I feel about this. Reflection is always a last resort, but at least before this PR it was only `protected` fields, part of the exposed API.

After this PR, reflection is used on a _compiler-generated closure_ type, with a field matching a variable name... This seems brittle, but I don't think we have much of a choice.

For example, this won't work in NUnit 4 (which we don't officially support yet in `Avalonia.Headless`), since the code has changed: https://github.com/nunit/nunit/blob/main/src/NUnitFramework/framework/Internal/Commands/SetUpTearDownCommand.cs

cc @maxkatz6

Reply:
For the record, I agree completely, this was just the best solution I could come up with to ensure these methods run. To be clear, there are workarounds users can implement; because asynchronous test execution works, one can forego the `[SetUp]`/`[TearDown]` attributes and manually invoke the methods from each test that needs them (e.g. [what we do here](https://github.com/haroohie-club/SerialLoops/pull/476/files#diff-d5c22ef6f456b049aa4d2f515ce6ef1423af9b358237ba96f5428d9504f4f888R31-R61)). This is not ideal, but if documented could certainly be adequate if this solution is determined to be too brittle.

The ideal solution is of course an NUnit change to make the test execution environment more extensible. Hopefully the issue you commented on can get prioritized to allow for that.

Reply:
Does NUnit 4 has better extensibility support? If it does, we can completely drop 3.x support together with reflection.